### ISSUE TITLE
[Feature Request] Are there plans to publish the snapshot crate?

### ISSUE THREAD

**Issue by bergwolf** [firecracker-microvm/firecracker#4162]
Hello firecracker maintainers,

First of all, thank you for creating such a wonderful project and sharing it with the rest of the world with open source. There are a lot of rust projects benefiting from firecracker and we are one of them :)

I'm writing in the hat of maintainers of the [`fuse-backend-rs`](https://github.com/cloud-hypervisor/fuse-backend-rs/pull/149) project. It is a library used by the [Nydus image service](https://nydus.dev/) to provide FUSE file system support on both Linux and MacOS. Recently we've been adding [hot-upgrade ability](https://github.com/cloud-hypervisor/fuse-backend-rs/pull/149) to the library, which heavily depends on firecracker's [`snapshot`](https://github.com/firecracker-microvm/firecracker/tree/main/src/snapshot) crate.

Since we do publish `fuse-backend-rs` on crates.io (https://crates.io/crates/fuse-backend-rs), we cannot really have a dependency on unpublished crates. Therefore, we'd love to seek advice. Are there plans to publish the snapshot crate on crate.io at some point? Or do you think it is better for us to just fork that port of code as it is not supposed to be maintained as a public library at the moment? If you have any other suggestions/thoughts, please do share them. Thanks a lot!

Cheers,
Tao

**Comment by bchalios** [comment]
Hi @bergwolf,

Thanks for getting in touch, it is always nice to see people using (parts of) Firecracker :)

We discussed this internally and we decided that we would not like to publish the `snapshot` crate in crates.io at the moment. Doing so, would mean that we need to take a commitment to maintain it in a way that's more complicated for us. We would need for example to be careful with changes in its public API so that we do not break downstream projects like yours. At the moment, this would slow us down considerably.

The snapshot feature is still in Developer preview and we are still trying to consolidate API and functionality in something that makes sense for the particular case of Firecracker. For example, we have announced changes in https://github.com/firecracker-microvm/firecracker/discussions/4137 that are related with the versioning and parts of the functionality. 
TLDR; the biggest changes are:
1. We will decouple Firecracker versions from snapshot versions
2. We will stop using the `versionize` crate for (de)serializing state
3. We will not support creating snapshots for older versions of the snapshot format.

Given that, I believe that the best way forward for you is that you fork the crate and continue development in that fork.

I hope that makes sense for you.

**Comment by bergwolf** [comment]
Hi @bchalios, 

Thanks for the promote response. We surely understand your decision and will go ahead to fork the crate and maintain it separately. And your changes in https://github.com/firecracker-microvm/firecracker/discussions/4137 make a lot of sense and we would benefit from it as well when designing the Nydus hot-upgrade strategy. Thanks!

Cheers,
Tao

**Comment by bchalios** [comment]
Thanks Tao,

Then I will go ahead and resolve this. Please feel free to get in touch if you have more questions regarding the snapshotting story (or anything else).

Cheers,
Babis
